WebアプリからのDB利用

takao0206 opened the pull request:
GitHubへのファイルのアップロード、ダウンロードのテストです。

takao0206 committed:
Postgresに接続する機能を作成。

takao0206 committed:
データの保存先をJSONからPostgresに変更。

takao0206 committed:
README.mdの修正。

takao0206 commented:
メモアプリのデータ保存先をデータベースに変更しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu commented on memo_app.rb:
```diff
@@ -3,22 +3,88 @@
+end
+
+def create_table_if_not_exists
+  with_connection do |connection|
+    table_exists = connection.exec(
+      "SELECT EXISTS (
+        SELECT FROM information_schema.tables WHERE table_name = '#{TABLE_NAME}'
+      );"
+    ).first['exists'] == 't'
+
+    unless table_exists
+      connection.exec(
+        "CREATE TABLE #{TABLE_NAME} (
+          id SERIAL PRIMARY KEY,
+          #{COLUMN_TITLE} VARCHAR(100) NOT NULL,
+          #{COLUMN_CONTENT} TEXT NOT NULL
+        );"
+      )
+    end
```
このあたり、 `CREATE TABLE IF NOT EXISTS` の構文を使えば簡単かなと思います。
https://www.postgresql.jp/document/16/html/sql-createtable.html#SQL-CREATETABLE-PARMS-IF-NOT-EXISTS

takao0206 replied:
SQLのコマンドのご紹介、ありがとうございます。
`CREATE TABLE IF NOT EXISTS`を使って、修正したいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -27,9 +93,8 @@ def find_memo(memos, id)
 
 get '/' do
-  save_memos([]) unless File.exist?(MEMOS_PATH)
+  load_memos
 
-  @memos = load_memos
   erb :index
 end
```
前回指摘漏れていましたが、一覧ページもRESTに沿ったPathにしましょう。
その上で、 `/` からリダイレクトしてあげると親切だと思います。

takao0206 replied:
ご指摘、ありがとうございます。
まだRESTに沿ったPath設計に不慣れなため、勉強になりました。
`get '/' `は一覧を表示しますので、`get '/memos'`に修正したいと思います。
また、リダイレクト先も`redirect '/memos'`にしたいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -3,22 +3,88 @@
+  connection = connect_db
+  yield(connection)
+ensure
+  connection&.close
```
`close` も考慮できていてよいですね :+1:

takao0206 replied:
ありがとうございます。
`close`は忘れやすいので、メソッドにしてみました。

yoshitsugu commented on memo_app.rb:
```diff
@@ -3,22 +3,88 @@
-  JSON.parse(File.read(MEMOS_PATH))
+  with_connection do |connection|
+    @memos = connection.exec("SELECT * FROM #{TABLE_NAME} ORDER BY id ASC;").map do |memo|
+      memo['title'].force_encoding('UTF-8')
+      memo['content'].force_encoding('UTF-8')
```
設定として `client_encoding` を指定するなどで全体としてUTF-8で正しく扱えるようにしたほうがよさそうです。今後もし memo 以外のロード部分を作ることになったときに都度 `force_encoding` しないといけない、というのはうっかり忘れそうなので、バグに つながりそうです。

takao0206 replied:
アドバイス、ありがとうございます。少し根の深い問題でした。PostgreSQLの良い勉強になりました。
- 今回のアドバイスを受け、データベースの設定から見直しました。すると、データベースのエンコードが`SQL_ASCII`になっていることが分かりました。
- 私がREADMEで書いた`CREATE DATABASE memo_app WITH OWNER memo_app_user;`でデータベースを作ると、エンコードがASCIIになるようです。実際に、`SELECT pg_encoding_to_char(encoding) FROM pg_database WHERE datname = 'memo_app';`を実行すると、エンコードが`SQL_ASCII`になっていました。
- そこで、データベースを作成するSQLを以下に変更しました（READMEも修正しました）。これで、データベースのエンコードがUTF8になると思います。
```sql
CREATE DATABASE memo_app WITH OWNER memo_app_user ENCODING 'UTF8' TEMPLATE template0;
```
- さらにデータベースに入ってくるデータのエンコードをUTF8にするために、Rubyのコードの中に、以下のclient_encodingのSQLを追加しました。
```ruby
connection.exec("SET client_encoding TO 'UTF8'")
```

**もしデータベース`memo_app`のエンコードがASCIIの場合、お手数をお掛けし申し訳ございませんが、データベースを作り直すところからお願いできればと思います。**

yoshitsugu commented on memo_app.rb:
```diff
@@ -38,23 +103,13 @@ def find_memo(memos, id)
 get '/memos/:id' do
-  memos = load_memos
-  @memo = find_memo(memos, params[:id])
+  @memo = find_memo(load_memos, params[:id])
```
このプラクティスではさほど影響ないと思いますが、 `find_memo` もSQL実行時点から `LIMIT 1` をつけて専用のものにしてあげたほうが実行が速くなるかと思います。

takao0206 replied:
なるほどです。SQLで一度全部検索したものに対してRubyで`find`するよりも、SQLで検索するほうが高速になるということですね。おっしゃる通りと思います。
SQLを用いて`find_memo`メソッドを修正したいと思います。

yoshitsugu reviewed (commented):
コメントしました！

takao0206 committed:
ご指摘事項の修正（4点）

- データベースのテーブルの有無をチェックする部分を、`CREATE TABLE IF NOT EXISTS`を使って、リファクタリングした。
- 一覧ページの表示のPathをRESTに沿ったPathに変更した。
- エンコードエラーを回避するため、データベースのエンコードをUTF8に変更し、クライアントのエンコードがUTF8になるようにSQLを追加した。
- `find_memo`メソッドを高速化した。

takao0206 commented:
アドバイス、ありがとうございました。
ご指摘いただきましたところを修正して、push致しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu reviewed (approved):
OKです！
